File: node_version/providers/registry.ts
import { loadConfig } from "../config.js"
import { LLMProvider, LLMProviderError } from "./base.js"

import { GeminiProvider } from "./gemini.js"
import { OpenAIProvider } from "./openai.js"
import { OllamaProvider } from "./ollama.js"

type ProviderConstructor = new (config?: any) => LLMProvider

const PROVIDER_REGISTRY: Record<string, ProviderConstructor> = {
  gemini: GeminiProvider,
  openai: OpenAIProvider,
  ollama: OllamaProvider,
}

export function listProviders(): string[] {
  return Object.keys(PROVIDER_REGISTRY)
}

export function getProvider(name: string): LLMProvider {

  const providerName = name.toLowerCase()

  const Provider = PROVIDER_REGISTRY[providerName]

  if (!Provider) {
    throw new LLMProviderError(`Unknown LLM provider '${providerName}'`)
  }

  const config = loadConfig()

  const providerConfig =
    config?.providers?.[providerName] ?? {}

  return new Provider(providerConfig)
}

export function getActiveProvider(
  override?: string
): LLMProvider {

  if (override) {
    return getProvider(override)
  }

  const config = loadConfig()

  const defaultProvider =
    config?.llm?.provider

  if (!defaultProvider) {
    throw new LLMProviderError(
      "No LLM provider configured. Run 'explainthisrepo init'."
    )
  }

  return getProvider(defaultProvider)
}

File: explain_this_repo/providers/registry.py
from __future__ import annotations

from typing import Dict, Type

from explain_this_repo.config import load_config
from explain_this_repo.providers.base import LLMProvider, LLMProviderError

_PROVIDER_REGISTRY: Dict[str, str] = {
    "gemini": "explain_this_repo.providers.gemini.GeminiProvider",
    "openai": "explain_this_repo.providers.openai.OpenAIProvider",
    "ollama": "explain_this_repo.providers.ollama.OllamaProvider",
    "anthropic": "explain_this_repo.providers.anthropic.AnthropicProvider",
    "groq": "explain_this_repo.providers.groq.GroqProvider",
    "openrouter": "explain_this_repo.providers.openrouter.OpenRouterProvider",
}


def list_providers() -> list[str]:
    return list(_PROVIDER_REGISTRY.keys())


def _import_provider_class(path: str) -> Type[LLMProvider]:
    module_path, class_name = path.rsplit(".", 1)
    module = __import__(module_path, fromlist=[class_name])
    provider_cls = getattr(module, class_name)
    return provider_cls


def get_provider(name: str) -> LLMProvider:
    name = name.lower()

    if name not in _PROVIDER_REGISTRY:
        raise LLMProviderError(
            f"Unknown LLM provider '{name}'. "
            f"Available providers: {', '.join(sorted(_PROVIDER_REGISTRY.keys()))}"
        )

    config = load_config() or {}

    provider_config = config.get("providers", {}).get(name, {})

    provider_path = _PROVIDER_REGISTRY[name]
    provider_cls = _import_provider_class(provider_path)

    return provider_cls(provider_config)


def get_active_provider(
    override: str | None = None,
) -> LLMProvider:

    if override:
        return get_provider(override)

    config = load_config() or {}

    default_provider = config.get("llm", {}).get("provider")

    if not default_provider:
        raise LLMProviderError(
            "No LLM provider configured.\n"
            "Run `explainthisrepo init` to configure a provider."
        )

    return get_provider(default_provider)
